From b6b8982aff1d72f4cbbe8250aa693dde17032d0f Mon Sep 17 00:00:00 2001
From: Luciano Pacheco <lucmult@chromium.org>
Date: Tue, 26 Apr 2022 04:38:16 +0000
Subject: [PATCH] M101: Prevent multiple SelectFileDialog in
 PasswordManagerPorter

Similar to crrev.com/c/3546821

(cherry picked from commit 72706a05c4b23808a426b1bd2f7db337735ad7ac)

Bug: 1309583
Change-Id: Ibf9e602fc033e16a505dd834ad07bf305b0ed725
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3568932
Auto-Submit: Luciano Pacheco <lucmult@chromium.org>
Reviewed-by: Vasilii Sukhanov <vasilii@chromium.org>
Commit-Queue: Vasilii Sukhanov <vasilii@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#989386}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3585317
Commit-Queue: Luciano Pacheco <lucmult@chromium.org>
Cr-Commit-Position: refs/branch-heads/4951@{#1060}
Cr-Branched-From: 27de6227ca357da0d57ae2c7b18da170c4651438-refs/heads/main@{#982481}
---
 .../ui/passwords/settings/password_manager_porter.cc      | 8 ++++++++
 1 file changed, 8 insertions(+)

--- a/chrome/browser/ui/passwords/settings/password_manager_porter.cc
+++ b/chrome/browser/ui/passwords/settings/password_manager_porter.cc
@@ -181,6 +181,10 @@
 // This method should never be called on Android (as there is no file selector),
 // and the relevant IDS constants are not present for Android.
 #if !BUILDFLAG(IS_ANDROID)
+  // Early return if the select file dialog is already active.
+  if (select_file_dialog_)
+    return;
+
   DCHECK(web_contents);
   profile_ = Profile::FromBrowserContext(web_contents->GetBrowserContext());
 
@@ -231,12 +235,16 @@
       ExportPasswordsToPath(path);
       break;
   }
+
+  select_file_dialog_.reset();
 }
 
 void PasswordManagerPorter::FileSelectionCanceled(void* params) {
   if (reinterpret_cast<uintptr_t>(params) == PASSWORD_EXPORT) {
     exporter_->Cancel();
   }
+
+  select_file_dialog_.reset();
 }
 
 void PasswordManagerPorter::ImportPasswordsFromPath(
